compiler/src/dotty/tools/dotc/transform/SetRootTree.scala

```diff
@@ -0,0 +1,45 @@
+package dotty.tools.dotc.transform
+
+import dotty.tools.dotc.CompilationUnit
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.transform.MegaPhase.MiniPhase
+
+/** Set the `rootTreeOrProvider` property of class symbols. */
+class SetRootTree extends MiniPhase {
```

Comment:
To traverse top down you should use a `MacroTransform` instead of `MiniPhase`. Look at `PostTyper`

https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/transform/PostTyper.scala#L63-L64

Reply:
I wanted this to be a mini phase to avoid introducing a whole new traversal, but it seems unavoidable given that this phase has to run right before pickling (there are no miniphases in `frontendPhases`).

This phase doesn't do any transformation of the trees, so it looks like it could be a simple `Phase`, or am I missing something?

Reply:
As I said below, this is not a full traversal and hence it should be quite fast. Linear with the depth of the package assuming you do a top-down traversal and stop when you find the top level class definitions.